streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java

```diff
@@ -192,13 +198,50 @@ private void encodePartitionsByHost(final DataOutputStream out) throws IOExcepti
         // encode partitions by host
         out.writeInt(partitionsByHost.size());
         for (final Map.Entry<HostInfo, Set<TopicPartition>> entry : partitionsByHost.entrySet()) {
-            final HostInfo hostInfo = entry.getKey();
-            out.writeUTF(hostInfo.host());
-            out.writeInt(hostInfo.port());
+            writeHostInfo(out, entry.getKey());
             writeTopicPartitions(out, entry.getValue());
         }
     }
 
+    private void encodePartitionsByHostAsDictionary(final DataOutputStream out) throws IOException {
```

Comment:
@guozhangwang Upon further testing, a simple dictionary encoding on the wire, saves almost the same amount as the index position based approach.. 

This is much much simpler and does not interfere with the protocol or topology changes in anyway. thoughts?